doc/overlays.xml

```diff
@@ -88,35 +88,37 @@ as <filename>overlays.nix</filename> and imported as the value of <literal>nixpk
 <title>Defining overlays</title>
 
 <para>Overlays are Nix functions which accept two arguments, 
-conventionally called <varname>self</varname> and <varname>super</varname>, 
+conventionally called <varname>final</varname> and <varname>previous</varname>,
 and return a set of packages. For example, the following is a valid overlay.</para>
 
 <programlisting>
-self: super:
+final: previous:
 
 {
-  boost = super.boost.override {
-    python = self.python3;
+  boost = previous.boost.override {
+    python = final.python3;
   };
-  rr = super.callPackage ./pkgs/rr {
-    stdenv = self.stdenv_32bit;
+  rr = final.callPackage ./pkgs/rr {
```

Comment:
`callPackage` is a function, which it-self captures `self` to provide the packages to the package, and using it through `self` will go through the fix-point twice.

Reply:
This does not make sense to me. For the interpreter the fix point is not a loop, and there is no such thing as the count of goes through the fix point. (Here is an accurate enough model of the internal states of a straightforward interpreter of the nix language that illustrates this: https://gist.github.com/orivej/5fddcf792dfa37a5bfa16d9a95fe70fa .) Using `final` is in no way less computationally efficient than using `previous`. When `callPackage` is not overridden in a future overlay, `final.callPackage` is exactly the same as `previous.callPackage`, but the choice of the former is more obvious.

Reply:
@orivej Today there is not, but the reason the documentation is that way, is to make sure that we make the code written by user future-proof.  Doing the opposite would become a problem for automatic grafting.

As you mentioned, there is no difference today, but doing it as currently mentioned in the documentation will prevent the addition of issues that would have to be fixed later.

Reply:
> I disagree, `callPackage` is a function, and it should come from `previous`, not `final`.

Could you explain why functions should come from `previous`? I am fairly certain that in practice this is a stylistic choice until someone wants to override some functions (and `final` looks better), and it should not concern us if the referenced value is a derivation or a function . The only concern is if we want to use the previous value to implement our override of it, or if we don't need the previous value.

Reply:
@nbp Could you detail the future plans? What is automatic grafting?

Reply:
From the name of it I guess the automatic grafting is deep merging of attribute sets, but that does not seem to cause any issues with using functions from `final`.

Reply:
@orivej The automatic grafting described in https://github.com/NixOS/nixpkgs/pull/10851 , which is coming slowly, depends on the fact that we can peel off the fix-point and watch the number of hops through the fix-point.  Using a function which captures `self`, from `self`, implies that we go twice through self, and miss this dependency.